// @ckb-ccc/joy-id/dist/evm/index.js
import { ccc } from "@ckb-ccc/core";
import { DappRequestType, buildJoyIDURL } from "@joyid/common";
import { createPopup } from "../common/index.js";
import { ConnectionsRepoLocalStorage, } from "../connectionsStorage/index.js";
/**
 * Class representing an EVM signer that extends SignerEvm
 * @public
 */
export class EvmSigner extends ccc.SignerEvm {
    /**
     * Ensures that the signer is connected and returns the connection.
     * @throws Will throw an error if not connected.
     * @returns The current connection.
     */
    async assertConnection() {
        if (!(await this.isConnected()) || !this.connection) {
            throw new Error("Not connected");
        }
        return this.connection;
    }
    /**
     * Creates an instance of EvmSigner.
     * @param client - The client instance.
     * @param name - The name of the signer.
     * @param icon - The icon URL of the signer.
     * @param _appUri - The application URI.
     * @param connectionsRepo - The connections repository.
     */
    constructor(client, name, icon, _appUri, connectionsRepo = new ConnectionsRepoLocalStorage()) {
        super(client);
        this.name = name;
        this.icon = icon;
        this._appUri = _appUri;
        this.connectionsRepo = connectionsRepo;
    }
    /**
     * Gets the configuration for JoyID.
     * @returns The configuration object.
     */
    getConfig() {
        return {
            redirectURL: location.href,
            joyidAppURL: this._appUri ??
                (this.client.addressPrefix === "ckb"
                    ? "https://app.joy.id"
                    : "https://testnet.joyid.dev"),
            requestNetwork: `ethereum`,
            name: this.name,
            logo: this.icon,
        };
    }
    /**
     * Gets the EVM account address.
     * @returns A promise that resolves to the EVM account address.
     */
    async getEvmAccount() {
        return (await this.assertConnection()).address;
    }
    /**
     * Connects to the provider by requesting authentication.
     * @returns A promise that resolves when the connection is established.
     */
    async connect() {
        const config = this.getConfig();
        const res = await createPopup(buildJoyIDURL(config, "popup", "/auth"), {
            ...config,
            type: DappRequestType.Auth,
        });
        this.connection = {
            address: res.ethAddress,
            publicKey: ccc.hexFrom(res.pubkey),
            keyType: res.keyType,
        };
        await this.saveConnection();
    }
    async disconnect() {
        await super.disconnect();
        this.connection = undefined;
        await this.saveConnection();
    }
    /**
     * Checks if the signer is connected.
     * @returns A promise that resolves to true if connected, false otherwise.
     */
    async isConnected() {
        if (this.connection) {
            return true;
        }
        await this.restoreConnection();
        return this.connection !== undefined;
    }
    /**
     * Signs a raw message with the EVM account.
     * @param message - The message to sign.
     * @returns A promise that resolves to the signed message.
     */
    async signMessageRaw(message) {
        const { address } = await this.assertConnection();
        const challenge = typeof message === "string" ? message : ccc.hexFrom(message).slice(2);
        const config = this.getConfig();
        const { signature } = await createPopup(buildJoyIDURL({
            ...config,
            challenge,
            isData: typeof message !== "string",
            address,
        }, "popup", "/sign-message"), { ...config, type: DappRequestType.SignMessage });
        return ccc.hexFrom(signature);
    }
    /**
     * Saves the current connection.
     * @returns
     */
    async saveConnection() {
        return this.connectionsRepo.set({
            uri: this.getConfig().joyidAppURL,
            addressType: "ethereum",
        }, this.connection);
    }
    /**
     * Restores the previous connection.
     * @returns
     */
    async restoreConnection() {
        this.connection = await this.connectionsRepo.get({
            uri: this.getConfig().joyidAppURL,
            addressType: "ethereum",
        });
    }
}
