// @ckb-ccc/xverse/dist.commonjs/signer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Signer = void 0;
const core_1 = require("@ckb-ccc/core");
const v = __importStar(require("valibot"));
const advancedBarrel_js_1 = require("./advancedBarrel.js");
async function checkResponse(response) {
    const res = await response;
    if (v.is(advancedBarrel_js_1.rpcErrorResponseMessageSchema, res)) {
        // eslint-disable-next-line @typescript-eslint/only-throw-error
        throw res.error;
    }
    if (v.is(advancedBarrel_js_1.rpcSuccessResponseMessageSchema, res)) {
        return res.result;
    }
    // eslint-disable-next-line @typescript-eslint/only-throw-error
    throw {
        code: advancedBarrel_js_1.RpcErrorCode.INTERNAL_ERROR,
        message: "Received unknown response from provider.",
        data: res,
    };
}
/**
 * Class representing a Bitcoin signer that extends SignerBtc
 * @public
 */
class Signer extends core_1.ccc.SignerBtc {
    /**
     * Creates an instance of Signer.
     * @param client - The client instance.
     * @param provider - The provider instance.
     */
    constructor(client, provider, preferredNetworks = [
        {
            addressPrefix: "ckb",
            signerType: core_1.ccc.SignerType.BTC,
            network: "btc",
        },
        {
            addressPrefix: "ckt",
            signerType: core_1.ccc.SignerType.BTC,
            network: "btcTestnet",
        },
    ]) {
        super(client);
        this.provider = provider;
        this.preferredNetworks = preferredNetworks;
    }
    async assertAddress() {
        this.addressCache =
            this.addressCache ??
                (async () => {
                    if (!(await this.isConnected())) {
                        return;
                    }
                    return (await checkResponse(this.provider.request("getAddresses", {
                        purposes: [advancedBarrel_js_1.AddressPurpose.Payment],
                    }))).addresses[0];
                })();
        const address = await this.addressCache;
        if (address) {
            return address;
        }
        throw Error("Not connected");
    }
    /**
     * Gets the Bitcoin account address.
     * @returns A promise that resolves to the Bitcoin account address.
     */
    async getBtcAccount() {
        return (await this.assertAddress()).address;
    }
    /**
     * Gets the Bitcoin public key.
     * @returns A promise that resolves to the Bitcoin public key.
     */
    async getBtcPublicKey() {
        return core_1.ccc.hexFrom((await this.assertAddress()).publicKey);
    }
    /**
     * Connects to the provider by requesting accounts.
     * @returns A promise that resolves when the connection is established.
     */
    async connect() {
        if (await this.isConnected()) {
            return;
        }
        await checkResponse(this.provider.request("wallet_requestPermissions", undefined));
    }
    async disconnect() {
        this.addressCache = undefined;
    }
    onReplaced(listener) {
        const stop = [];
        const replacer = () => {
            listener();
            stop[0]?.();
        };
        stop.push(this.provider.addListener("accountChange", replacer), this.provider.addListener("networkChange", replacer));
        return stop[0];
    }
    /**
     * Checks if the signer is connected.
     * @returns A promise that resolves to true if connected, false otherwise.
     */
    async isConnected() {
        try {
            await checkResponse(this.provider.request("getBalance", undefined));
            return true;
        }
        catch (_error) {
            return false;
        }
    }
    /**
     * Signs a raw message with the Bitcoin account.
     * @param message - The message to sign.
     * @returns A promise that resolves to the signed message.
     */
    async signMessageRaw(message) {
        const challenge = typeof message === "string" ? message : core_1.ccc.hexFrom(message).slice(2);
        return (await checkResponse(this.provider.request("signMessage", {
            message: challenge,
            address: (await this.assertAddress()).address,
            protocol: advancedBarrel_js_1.MessageSigningProtocols.ECDSA,
        }))).signature;
    }
}
exports.Signer = Signer;
