// @ckb-ccc/xverse/dist/signer.js
import { ccc } from "@ckb-ccc/core";
import * as v from "valibot";
import { AddressPurpose, MessageSigningProtocols, RpcErrorCode, rpcErrorResponseMessageSchema, rpcSuccessResponseMessageSchema, } from "./advancedBarrel.js";
async function checkResponse(response) {
    const res = await response;
    if (v.is(rpcErrorResponseMessageSchema, res)) {
        // eslint-disable-next-line @typescript-eslint/only-throw-error
        throw res.error;
    }
    if (v.is(rpcSuccessResponseMessageSchema, res)) {
        return res.result;
    }
    // eslint-disable-next-line @typescript-eslint/only-throw-error
    throw {
        code: RpcErrorCode.INTERNAL_ERROR,
        message: "Received unknown response from provider.",
        data: res,
    };
}
/**
 * Class representing a Bitcoin signer that extends SignerBtc
 * @public
 */
export class Signer extends ccc.SignerBtc {
    /**
     * Creates an instance of Signer.
     * @param client - The client instance.
     * @param provider - The provider instance.
     */
    constructor(client, provider, preferredNetworks = [
        {
            addressPrefix: "ckb",
            signerType: ccc.SignerType.BTC,
            network: "btc",
        },
        {
            addressPrefix: "ckt",
            signerType: ccc.SignerType.BTC,
            network: "btcTestnet",
        },
    ]) {
        super(client);
        this.provider = provider;
        this.preferredNetworks = preferredNetworks;
    }
    async assertAddress() {
        this.addressCache =
            this.addressCache ??
                (async () => {
                    if (!(await this.isConnected())) {
                        return;
                    }
                    return (await checkResponse(this.provider.request("getAddresses", {
                        purposes: [AddressPurpose.Payment],
                    }))).addresses[0];
                })();
        const address = await this.addressCache;
        if (address) {
            return address;
        }
        throw Error("Not connected");
    }
    /**
     * Gets the Bitcoin account address.
     * @returns A promise that resolves to the Bitcoin account address.
     */
    async getBtcAccount() {
        return (await this.assertAddress()).address;
    }
    /**
     * Gets the Bitcoin public key.
     * @returns A promise that resolves to the Bitcoin public key.
     */
    async getBtcPublicKey() {
        return ccc.hexFrom((await this.assertAddress()).publicKey);
    }
    /**
     * Connects to the provider by requesting accounts.
     * @returns A promise that resolves when the connection is established.
     */
    async connect() {
        if (await this.isConnected()) {
            return;
        }
        await checkResponse(this.provider.request("wallet_requestPermissions", undefined));
    }
    async disconnect() {
        this.addressCache = undefined;
    }
    onReplaced(listener) {
        const stop = [];
        const replacer = () => {
            listener();
            stop[0]?.();
        };
        stop.push(this.provider.addListener("accountChange", replacer), this.provider.addListener("networkChange", replacer));
        return stop[0];
    }
    /**
     * Checks if the signer is connected.
     * @returns A promise that resolves to true if connected, false otherwise.
     */
    async isConnected() {
        try {
            await checkResponse(this.provider.request("getBalance", undefined));
            return true;
        }
        catch (_error) {
            return false;
        }
    }
    /**
     * Signs a raw message with the Bitcoin account.
     * @param message - The message to sign.
     * @returns A promise that resolves to the signed message.
     */
    async signMessageRaw(message) {
        const challenge = typeof message === "string" ? message : ccc.hexFrom(message).slice(2);
        return (await checkResponse(this.provider.request("signMessage", {
            message: challenge,
            address: (await this.assertAddress()).address,
            protocol: MessageSigningProtocols.ECDSA,
        }))).signature;
    }
}
