// @ckb-ccc/xverse/dist/sat-connect-core/btcMethods.advanced.js
// From https://github.com/secretkeylabs/sats-connect-core/
/**
 * Represents the types and interfaces related to BTC methods.
 */
import * as v from "valibot";
import { rpcRequestMessageSchema, } from "./types.advanced.js";
import { walletTypeSchema } from "./walletMethods.advanced.js";
export var AddressPurpose;
(function (AddressPurpose) {
    AddressPurpose["Ordinals"] = "ordinals";
    AddressPurpose["Payment"] = "payment";
    AddressPurpose["Stacks"] = "stacks";
})(AddressPurpose || (AddressPurpose = {}));
export var AddressType;
(function (AddressType) {
    AddressType["p2pkh"] = "p2pkh";
    AddressType["p2sh"] = "p2sh";
    AddressType["p2wpkh"] = "p2wpkh";
    AddressType["p2wsh"] = "p2wsh";
    AddressType["p2tr"] = "p2tr";
    AddressType["stacks"] = "stacks";
})(AddressType || (AddressType = {}));
export const addressSchema = v.object({
    address: v.string(),
    publicKey: v.string(),
    purpose: v.enum(AddressPurpose),
    addressType: v.enum(AddressType),
});
export const getInfoMethodName = "getInfo";
export const getInfoParamsSchema = v.nullish(v.null());
export const getInfoResultSchema = v.object({
    /**
     * Version of the wallet.
     */
    version: v.string(),
    /**
     * [WBIP](https://wbips.netlify.app/wbips/WBIP002) methods supported by the wallet.
     */
    methods: v.optional(v.array(v.string())),
    /**
     * List of WBIP standards supported by the wallet. Not currently used.
     */
    supports: v.array(v.string()),
});
export const getInfoRequestMessageSchema = v.object({
    ...rpcRequestMessageSchema.entries,
    ...v.object({
        method: v.literal(getInfoMethodName),
        params: getInfoParamsSchema,
        id: v.string(),
    }).entries,
});
export const getAddressesMethodName = "getAddresses";
export const getAddressesParamsSchema = v.object({
    /**
     * The purposes for which to generate addresses. See
     * {@linkcode AddressPurpose} for available purposes.
     */
    purposes: v.array(v.enum(AddressPurpose)),
    /**
     * A message to be displayed to the user in the request prompt.
     */
    message: v.optional(v.string()),
});
export const getAddressesResultSchema = v.object({
    /**
     * The addresses generated for the given purposes.
     */
    addresses: v.array(addressSchema),
});
export const getAddressesRequestMessageSchema = v.object({
    ...rpcRequestMessageSchema.entries,
    ...v.object({
        method: v.literal(getAddressesMethodName),
        params: getAddressesParamsSchema,
        id: v.string(),
    }).entries,
});
export const signMessageMethodName = "signMessage";
export var MessageSigningProtocols;
(function (MessageSigningProtocols) {
    MessageSigningProtocols["ECDSA"] = "ECDSA";
    MessageSigningProtocols["BIP322"] = "BIP322";
})(MessageSigningProtocols || (MessageSigningProtocols = {}));
export const signMessageParamsSchema = v.object({
    /**
     * The address used for signing.
     **/
    address: v.string(),
    /**
     * The message to sign.
     **/
    message: v.string(),
    /**
     * The protocol to use for signing the message.
     */
    protocol: v.optional(v.enum(MessageSigningProtocols)),
});
export const signMessageResultSchema = v.object({
    /**
     * The signature of the message.
     */
    signature: v.string(),
    /**
     * hash of the message.
     */
    messageHash: v.string(),
    /**
     * The address used for signing.
     */
    address: v.string(),
    /**
     * The protocol to use for signing the message.
     */
    protocol: v.enum(MessageSigningProtocols),
});
export const signMessageRequestMessageSchema = v.object({
    ...rpcRequestMessageSchema.entries,
    ...v.object({
        method: v.literal(signMessageMethodName),
        params: signMessageParamsSchema,
        id: v.string(),
    }).entries,
});
export const getAccountsMethodName = "getAccounts";
export const getAccountsParamsSchema = v.object({
    /**
     * The purposes for which to generate addresses. See
     * {@linkcode AddressPurpose} for available purposes.
     */
    purposes: v.array(v.enum(AddressPurpose)),
    /**
     * A message to be displayed to the user in the request prompt.
     */
    message: v.optional(v.string()),
});
export const getAccountsResultSchema = v.array(v.object({
    ...addressSchema.entries,
    ...v.object({
        walletType: walletTypeSchema,
    }).entries,
}));
export const getAccountsRequestMessageSchema = v.object({
    ...rpcRequestMessageSchema.entries,
    ...v.object({
        method: v.literal(getAccountsMethodName),
        params: getAccountsParamsSchema,
        id: v.string(),
    }).entries,
});
export const getBalanceMethodName = "getBalance";
export const getBalanceParamsSchema = v.nullish(v.null());
export const getBalanceResultSchema = v.object({
    /**
     * The confirmed balance of the wallet in sats. Using a string due to chrome
     * messages not supporting bigint
     * (https://issues.chromium.org/issues/40116184).
     */
    confirmed: v.string(),
    /**
     * The unconfirmed balance of the wallet in sats. Using a string due to chrome
     * messages not supporting bigint
     * (https://issues.chromium.org/issues/40116184).
     */
    unconfirmed: v.string(),
    /**
     * The total balance (both confirmed and unconfrimed UTXOs) of the wallet in
     * sats. Using a string due to chrome messages not supporting bigint
     * (https://issues.chromium.org/issues/40116184).
     */
    total: v.string(),
});
export const getBalanceRequestMessageSchema = v.object({
    ...rpcRequestMessageSchema.entries,
    ...v.object({
        method: v.literal(getBalanceMethodName),
        id: v.string(),
    }).entries,
});
