// @ckb-ccc/xverse/dist/sat-connect-core/types.advanced.js
/* eslint-disable @typescript-eslint/no-explicit-any */
// From https://github.com/secretkeylabs/sats-connect-core/
import * as v from "valibot";
export var BitcoinNetworkType;
(function (BitcoinNetworkType) {
    BitcoinNetworkType["Mainnet"] = "Mainnet";
    BitcoinNetworkType["Testnet"] = "Testnet";
    BitcoinNetworkType["Signet"] = "Signet";
})(BitcoinNetworkType || (BitcoinNetworkType = {}));
// RPC Request and Response types
export const RpcIdSchema = v.optional(v.union([v.string(), v.number(), v.null()]));
export const rpcRequestMessageSchema = v.object({
    jsonrpc: v.literal("2.0"),
    method: v.string(),
    params: v.optional(v.union([
        v.array(v.unknown()),
        v.looseObject({}),
        // Note: This is to support current incorrect usage of RPC 2.0. Params need
        // to be either an array or an object when provided. Changing this now would
        // be a breaking change, so accepting null values for now. Tracking in
        // https://linear.app/xverseapp/issue/ENG-4538.
        v.null(),
    ])),
    id: RpcIdSchema,
});
/**
 * @enum {number} RpcErrorCode
 * @description JSON-RPC error codes
 * @see https://www.jsonrpc.org/specification#error_object
 */
export var RpcErrorCode;
(function (RpcErrorCode) {
    /**
     * Parse error Invalid JSON
     **/
    RpcErrorCode[RpcErrorCode["PARSE_ERROR"] = -32700] = "PARSE_ERROR";
    /**
     * The JSON sent is not a valid Request object.
     **/
    RpcErrorCode[RpcErrorCode["INVALID_REQUEST"] = -32600] = "INVALID_REQUEST";
    /**
     * The method does not exist/is not available.
     **/
    RpcErrorCode[RpcErrorCode["METHOD_NOT_FOUND"] = -32601] = "METHOD_NOT_FOUND";
    /**
     * Invalid method parameter(s).
     */
    RpcErrorCode[RpcErrorCode["INVALID_PARAMS"] = -32602] = "INVALID_PARAMS";
    /**
     * Internal JSON-RPC error.
     * This is a generic error, used when the server encounters an error in performing the request.
     **/
    RpcErrorCode[RpcErrorCode["INTERNAL_ERROR"] = -32603] = "INTERNAL_ERROR";
    /**
     * user rejected/canceled the request
     */
    RpcErrorCode[RpcErrorCode["USER_REJECTION"] = -32000] = "USER_REJECTION";
    /**
     * method is not supported for the address provided
     */
    RpcErrorCode[RpcErrorCode["METHOD_NOT_SUPPORTED"] = -32001] = "METHOD_NOT_SUPPORTED";
    /**
     * The client does not have permission to access the requested resource.
     */
    RpcErrorCode[RpcErrorCode["ACCESS_DENIED"] = -32002] = "ACCESS_DENIED";
})(RpcErrorCode || (RpcErrorCode = {}));
export const rpcSuccessResponseMessageSchema = v.object({
    jsonrpc: v.literal("2.0"),
    result: v.nonOptional(v.unknown()),
    id: RpcIdSchema,
});
export const rpcErrorResponseMessageSchema = v.object({
    jsonrpc: v.literal("2.0"),
    error: v.nonOptional(v.unknown()),
    id: RpcIdSchema,
});
export const rpcResponseMessageSchema = v.union([
    rpcSuccessResponseMessageSchema,
    rpcErrorResponseMessageSchema,
]);
