// @ckb-ccc/xverse/dist/sat-connect-core/walletMethods.advanced.js
// From https://github.com/secretkeylabs/sats-connect-core/
import * as v from "valibot";
import { rpcRequestMessageSchema, } from "./types.advanced.js";
export const walletTypes = ["software", "ledger"];
export const walletTypeSchema = v.picklist(walletTypes);
export const requestPermissionsMethodName = "wallet_requestPermissions";
export const requestPermissionsParamsSchema = v.undefined();
export const requestPermissionsResultSchema = v.literal(true);
export const requestPermissionsRequestMessageSchema = v.object({
    ...rpcRequestMessageSchema.entries,
    ...v.object({
        method: v.literal(requestPermissionsMethodName),
        params: requestPermissionsParamsSchema,
        id: v.string(),
    }).entries,
});
export const renouncePermissionsMethodName = "wallet_renouncePermissions";
export const renouncePermissionsParamsSchema = v.undefined();
export const renouncePermissionsResultSchema = v.literal(true);
export const renouncePermissionsRequestMessageSchema = v.object({
    ...rpcRequestMessageSchema.entries,
    ...v.object({
        method: v.literal(renouncePermissionsMethodName),
        params: renouncePermissionsParamsSchema,
        id: v.string(),
    }).entries,
});
export const getWalletTypeMethodName = "wallet_getWalletType";
export const getWalletTypeParamsSchema = v.nullish(v.null());
export const getWalletTypeResultSchema = walletTypeSchema;
export const getWalletTypeRequestMessageSchema = v.object({
    ...rpcRequestMessageSchema.entries,
    ...v.object({
        method: v.literal(getWalletTypeMethodName),
        id: v.string(),
    }).entries,
});
