// @ckb-ccc/spore/dist/codec/cluster.js
import { ccc, mol } from "@ckb-ccc/core";
export const ClusterDataV1 = mol.table({
    name: mol.String,
    description: mol.String,
});
export const ClusterDataV2 = mol.table({
    name: mol.String,
    description: mol.String,
    mutantId: mol.BytesOpt,
});
export const ClusterData = ClusterDataV2;
export function packRawClusterData(packable, version) {
    if (!version) {
        return packRawClusterDataV2(packable);
    }
    switch (version) {
        case "v1":
            return packRawClusterDataV1(packable);
        case "v2":
            return packRawClusterDataV2(packable);
    }
}
export function packRawClusterDataV1(packable) {
    return ccc.bytesFrom(ClusterDataV1.encode({
        name: packable.name,
        description: packable.description,
    }));
}
export function packRawClusterDataV2(packable) {
    return ccc.bytesFrom(ClusterDataV2.encode({
        name: packable.name,
        description: packable.description,
        mutantId: packable.mutantId,
    }));
}
export function unpackToRawClusterData(unpackable, version) {
    if (version) {
        switch (version) {
            case "v1":
                return unpackToRawClusterDataV1(unpackable);
            case "v2":
                return unpackToRawClusterDataV2(unpackable);
        }
    }
    try {
        return unpackToRawClusterDataV2(unpackable);
    }
    catch {
        try {
            return unpackToRawClusterDataV1(unpackable);
        }
        catch {
            throw new Error(`Cannot unpack ClusterData, no matching molecule: ${ccc.hexFrom(unpackable)}`);
        }
    }
}
export function unpackToRawClusterDataV1(unpackable) {
    return ClusterDataV1.decode(unpackable);
}
export function unpackToRawClusterDataV2(unpackable) {
    return ClusterDataV2.decode(unpackable);
}
